File: desktop/nativeMediaCoreStateMapper.ts
/**
 * Maps the compact wire state returned by `corevideo-native.exe` into the full
 * `NativeMediaCoreStateSnapshot` shape the renderer expects. The Node stub core
 * already emits the full snapshot; this adapter keeps the C++ core slim while
 * preserving the typed boundary contract.
 */
import type {
  NativeMediaCoreCommand,
  NativeMediaCoreEncoderSession,
  NativeMediaCoreOutputHealth,
  NativeMediaCoreOutputSenderSession,
  NativeMediaCoreProfile,
  NativeMediaCoreRecordingSession,
  NativeMediaCoreStateSnapshot
} from "../src/engine/nativeMediaCoreProtocol";
import { synthesizeSnapshot } from "./syntheticMediaCore.ts";

export type NativeMediaCoreWireState = {
  sceneId?: string;
  routeCount?: number;
  transformCount?: number;
  overlayCount?: number;
  outputs?: string[];
  isoParticipantIds?: string[];
  encoder?: string;
  codec?: string;
  hardwareEncoder?: boolean;
  active?: boolean;
  programFrameCount?: number;
  renderPlanId?: string;
  compositorRenderer?: string;
  encoderSession?: NativeMediaCoreEncoderSession;
  outputSenderSession?: NativeMediaCoreOutputSenderSession;
  recording?: NativeMediaCoreRecordingSession;
  health?: {
    status?: string;
    renderer?: string;
    encoder?: string;
    codec?: string;
    hardwareEncoder?: boolean;
    recordingArtifactPath?: string;
    recordingBytesWritten?: number;
    encodedFrameCount?: number;
    frameCount?: number;
    messages?: string[];
  };
  profile?: NativeMediaCoreProfile;
};

function asDestination(value: string): "rtmp" | "ndi" | "srt" | "webrtc" | "recording" | undefined {
  if (value === "rtmp" || value === "ndi" || value === "srt" || value === "webrtc" || value === "recording") {
    return value;
  }
  return undefined;
}

function buildOutputHealth(
  outputs: Array<"rtmp" | "ndi" | "srt" | "webrtc" | "recording">,
  senderSession: NativeMediaCoreOutputSenderSession | undefined,
  recording: NativeMediaCoreRecordingSession | undefined,
  encoderWarnings: string[]
): NativeMediaCoreOutputHealth[] {
  const health: NativeMediaCoreOutputHealth[] = [];

  for (const destination of outputs) {
    if (destination === "recording") {
      const status =
        recording?.status === "failed"
          ? "failed"
          : recording?.status === "warning" || recording?.writerStatus === "warning"
            ? "warning"
            : recording?.active
              ? "live"
              : "idle";
      health.push({
        destination,
        status,
        message:
          recording?.error ?? recording?.warning ?? recording?.programPath ?? "Recording idle.",
        droppedFrames: recording?.totalDroppedFrames ?? 0
      });
      continue;
    }

    const sender = senderSession?.senders.find((item) => item.destination === destination);
    health.push({
      destination,
      status:
        sender?.status === "failed"
          ? "failed"
          : sender?.status === "warning"
            ? "warning"
            : sender?.status === "live" || sender?.status === "starting"
              ? "live"
              : "idle",
      message: sender?.warning ?? `${destination.toUpperCase()} sender ${sender?.status ?? "idle"}.`,
      droppedFrames: 0
    });
  }

  encoderWarnings.forEach((warning) => {
    health.push({
      destination: "recording",
      status: "warning",
      message: warning,
      droppedFrames: 0
    });
  });

  return health;
}

export function mapNativeWireStateToSnapshot(
  commands: NativeMediaCoreCommand[],
  elapsedMs: number,
  frameNumber: number,
  wire: NativeMediaCoreWireState
): NativeMediaCoreStateSnapshot {
  const base = synthesizeSnapshot(commands, elapsedMs, frameNumber);
  const outputs = (wire.outputs ?? base.outputs).map(asDestination).filter((value): value is NonNullable<typeof value> => Boolean(value));
  const programFrameCount = Math.max(
    wire.programFrameCount ?? 0,
    wire.health?.frameCount ?? 0,
    wire.encoderSession?.programFrameCount ?? 0,
    base.programFrameCount
  );
  const encoderSession = wire.encoderSession ?? base.encoderSession;
  const outputSenderSession = wire.outputSenderSession ?? base.outputSenderSession;
  const recording = wire.recording
    ? {
        ...wire.recording,
        totalBytesWritten: Math.max(wire.recording.totalBytesWritten, wire.health?.recordingBytesWritten ?? 0)
      }
    : wire.health?.recordingArtifactPath && base.recording
      ? {
          ...base.recording,
          totalBytesWritten: Math.max(base.recording.totalBytesWritten, wire.health.recordingBytesWritten ?? 0)
        }
      : base.recording;

  const compositorRenderer = wire.compositorRenderer ?? wire.health?.renderer ?? wire.profile?.renderer ?? "software";
  const compositorStatus =
    programFrameCount > 0 ? ("live" as const) : compositorRenderer === "software" ? base.compositor.status : ("idle" as const);
  const warnings = [...base.warnings];
  if (wire.health?.recordingArtifactPath && recording?.active) {
    warnings.unshift(`Recording artifact: ${wire.health.recordingArtifactPath}`);
  }
  if (wire.health?.messages?.length) {
    warnings.push(...wire.health.messages.filter((message) => message && !warnings.includes(message)));
  }

  const outputHealth = buildOutputHealth(outputs, outputSenderSession, recording, encoderSession.warnings ?? []);

  const programFrame = base.programFrame
    ? {
        ...base.programFrame,
        frameNumber: programFrameCount || base.programFrame.frameNumber,
        renderPlanId: wire.renderPlanId || base.programFrame.renderPlanId
      }
    : programFrameCount > 0
      ? {
          frameNumber: programFrameCount,
          timestampMs: elapsedMs,
          renderPlanId: wire.renderPlanId ?? `native-plan-${frameNumber}`,
          sceneId: wire.sceneId ?? base.sceneId,
          width: base.outputProfile.width,
          height: base.outputProfile.height,
          fps: base.outputProfile.fps,
          layerCount: base.renderPlan.layers.length,
          colorGrade: base.renderPlan.colorGrade,
          health: "live" as const
        }
      : undefined;

  const diagnostics = {
    ...base.diagnostics,
    generatedAtMs: elapsedMs,
    sceneId: wire.sceneId ?? base.sceneId,
    routeCount: wire.routeCount ?? base.routeCount,
    frameCount: base.frameCount,
    programFrameCount,
    outputs,
    outputHealth,
    outputSenderSession,
    compositor: {
      ...base.compositor,
      status: compositorStatus,
      renderPlanId: programFrame?.renderPlanId,
      programFrameCount,
      lastFrame: programFrame
    },
    programFrame,
    programTransport: {
      ...base.programTransport,
      status: programFrame ? ("publishing" as const) : ("idle" as const),
      frameNumber: programFrame?.frameNumber,
      renderPlanId: programFrame?.renderPlanId,
      timestampMs: programFrame?.timestampMs
    },
    encoderSession,
    recording,
    warnings
  };

  return {
    ...base,
    sceneId: wire.sceneId ?? base.sceneId,
    routeCount: wire.routeCount ?? base.routeCount,
    participantTransformCount: wire.transformCount ?? base.participantTransformCount,
    overlayCount: wire.overlayCount ?? base.overlayCount,
    outputs,
    isoParticipantIds: wire.isoParticipantIds ?? base.isoParticipantIds,
    programFrame,
    programFrameCount,
    compositor: diagnostics.compositor,
    programTransport: diagnostics.programTransport,
    encoderSession,
    outputSenderSession,
    outputHealth,
    recording,
    diagnostics,
    warnings
  };
}

export function isNativeMediaCoreWireState(value: unknown): value is NativeMediaCoreWireState {
  return typeof value === "object" && value !== null && "health" in value && "profile" in value && !("diagnostics" in value);
}

File: native-shell/CoreVideoPro.MediaCore/Services/NativeMediaCoreStateMapper.cs
using System.Text.Json;
using CoreVideoPro.MediaCore.Models;

namespace CoreVideoPro.MediaCore.Services;

public static class NativeMediaCoreStateMapper
{
    public static bool IsNativeMediaCoreWireState(JsonElement value)
    {
        return value.ValueKind == JsonValueKind.Object &&
               value.TryGetProperty("health", out _) &&
               value.TryGetProperty("profile", out _) &&
               !value.TryGetProperty("diagnostics", out _);
    }

    public static NativeMediaCoreStateSnapshot MapNativeWireStateToSnapshot(
        IReadOnlyList<NativeMediaCoreCommand> commands,
        double elapsedMs,
        int frameNumber,
        NativeMediaCoreWireState wire)
    {
        var baseSnapshot = SyntheticMediaCore.SynthesizeSnapshot(commands, elapsedMs, frameNumber);
        var outputs = (wire.Outputs ?? baseSnapshot.Outputs)
            .Select(AsDestination)
            .Where(destination => destination is not null)
            .Select(destination => destination!)
            .ToList();

        var programFrameCount = Math.Max(
            wire.ProgramFrameCount ?? 0,
            wire.Health?.FrameCount ?? 0);
        programFrameCount = Math.Max(programFrameCount, wire.EncoderSession?.ProgramFrameCount ?? 0);

        var encoderSession = wire.EncoderSession ?? baseSnapshot.EncoderSession;
        var outputSenderSession = wire.OutputSenderSession ?? baseSnapshot.OutputSenderSession;
        var recording = wire.Recording is not null
            ? wire.Recording with
            {
                TotalBytesWritten = Math.Max(
                    wire.Recording.TotalBytesWritten,
                    wire.Health?.RecordingBytesWritten ?? 0)
            }
            : wire.Health?.RecordingArtifactPath is not null && baseSnapshot.Recording is not null
                ? baseSnapshot.Recording with
                {
                    TotalBytesWritten = Math.Max(
                        baseSnapshot.Recording.TotalBytesWritten,
                        wire.Health?.RecordingBytesWritten ?? 0)
                }
                : baseSnapshot.Recording;

        var compositorRenderer = wire.CompositorRenderer ??
                                 wire.Health?.Renderer ??
                                 wire.Profile?.Renderer ??
                                 "software";
        var programFrameHealth = NormalizeProgramFrameHealth(wire.Health?.ProgramFrameHealth);
        var compositorStatus = programFrameCount > 0
            ? programFrameHealth
            : compositorRenderer.Equals("software", StringComparison.Ordinal)
                ? "idle"
                : "idle";

        var warnings = baseSnapshot.Warnings.ToList();
        if (wire.Health?.RecordingArtifactPath is not null && recording?.Active == true)
        {
            warnings.Insert(0, $"Recording artifact: {wire.Health.RecordingArtifactPath}");
        }

        if (wire.Health?.Messages is not null)
        {
            foreach (var message in wire.Health.Messages.Where(message => !string.IsNullOrWhiteSpace(message)))
            {
                if (!warnings.Contains(message))
                {
                    warnings.Add(message);
                }
            }
        }

        var outputHealth = BuildOutputHealth(outputs, outputSenderSession, recording, encoderSession.Warnings);
        var programFramePreview = wire.ProgramFramePreview is not null
            ? DecodeProgramFramePreview(wire.ProgramFramePreview) ?? baseSnapshot.ProgramFramePreview
            : baseSnapshot.ProgramFramePreview;

        var programFrame = programFrameCount > 0
            ? new NativeMediaCoreProgramFrame
            {
                FrameNumber = programFrameCount,
                TimestampMs = elapsedMs,
                RenderPlanId = wire.RenderPlanId ?? baseSnapshot.ProgramFrame?.RenderPlanId ?? $"native-plan-{frameNumber}",
                SceneId = wire.SceneId ?? baseSnapshot.SceneId,
                Width = baseSnapshot.OutputProfile.Width,
                Height = baseSnapshot.OutputProfile.Height,
                Fps = baseSnapshot.OutputProfile.Fps,
                LayerCount = baseSnapshot.RenderPlan.Layers.Count,
                ColorGrade = baseSnapshot.RenderPlan.ColorGrade,
                Health = programFrameHealth
            }
            : null;

        var audioMixSession = wire.AudioMixSession ?? baseSnapshot.AudioMixSession;
        var captionTrack = wire.CaptionTrack ?? baseSnapshot.CaptionTrack;
        var brandKit = wire.BrandKit ?? baseSnapshot.BrandKit;
        var mergedWarnings = warnings
            .Concat(audioMixSession.Warnings.Where(warning => !warnings.Contains(warning)))
            .Concat(captionTrack.Warnings.Where(warning => !warnings.Contains(warning)))
            .Concat(brandKit.Warnings.Where(warning => !warnings.Contains(warning)))
            .Distinct()
            .ToList();

        var compositor = baseSnapshot.Compositor with
        {
            Status = compositorStatus,
            RenderPlanId = programFrame?.RenderPlanId,
            ProgramFrameCount = programFrameCount,
            LastFrame = programFrame
        };

        var programTransport = baseSnapshot.ProgramTransport with
        {
            Status = programFrame is null ? "idle" : "publishing",
            FrameNumber = programFrame?.FrameNumber,
            RenderPlanId = programFrame?.RenderPlanId,
            TimestampMs = programFrame?.TimestampMs
        };

        var diagnostics = baseSnapshot.Diagnostics with
        {
            GeneratedAtMs = elapsedMs,
            SceneId = wire.SceneId ?? baseSnapshot.SceneId,
            RouteCount = wire.RouteCount ?? baseSnapshot.RouteCount,
            FrameCount = baseSnapshot.FrameCount,
            ProgramFrameCount = programFrameCount,
            Outputs = outputs,
            OutputHealth = outputHealth,
            OutputSenderSession = outputSenderSession,
            Compositor = compositor,
            ProgramFrame = programFrame,
            ProgramFramePreview = programFramePreview,
            ProgramTransport = programTransport,
            EncoderSession = encoderSession,
            Recording = recording,
            AudioMixSession = audioMixSession,
            CaptionTrack = captionTrack,
            BrandKit = brandKit,
            Warnings = mergedWarnings
        };

        return baseSnapshot with
        {
            SceneId = wire.SceneId ?? baseSnapshot.SceneId,
            RouteCount = wire.RouteCount ?? baseSnapshot.RouteCount,
            ParticipantTransformCount = wire.TransformCount ?? baseSnapshot.ParticipantTransformCount,
            OverlayCount = wire.OverlayCount ?? baseSnapshot.OverlayCount,
            Outputs = outputs,
            IsoParticipantIds = wire.IsoParticipantIds ?? baseSnapshot.IsoParticipantIds,
            ProgramFrame = programFrame,
            ProgramFramePreview = programFramePreview,
            ProgramFrameCount = programFrameCount,
            Compositor = compositor,
            ProgramTransport = programTransport,
            EncoderSession = encoderSession,
            OutputSenderSession = outputSenderSession,
            OutputHealth = outputHealth,
            Recording = recording,
            AudioMixSession = audioMixSession,
            CaptionTrack = captionTrack,
            BrandKit = brandKit,
            Diagnostics = diagnostics,
            Warnings = mergedWarnings,
            MeetingState = wire.MeetingState ?? baseSnapshot.MeetingState,
            BreakoutRoomId = wire.BreakoutRoomId ?? baseSnapshot.BreakoutRoomId,
            BreakoutRoomName = wire.BreakoutRoomName ?? baseSnapshot.BreakoutRoomName,
            ActiveSpeakerId = wire.ActiveSpeakerId ?? baseSnapshot.ActiveSpeakerId,
            Participants = wire.Participants is { Count: > 0 } participants
                ? participants
                : baseSnapshot.Participants
        };
    }

    private static string? AsDestination(string value) =>
        value is "rtmp" or "ndi" or "srt" or "webrtc" or "recording" ? value : null;

    private static string NormalizeProgramFrameHealth(string? health) =>
        health switch
        {
            "degraded" => "degraded",
            "dropped" => "failed",
            "failed" => "failed",
            "error" => "failed",
            _ => "live"
        };

    private static NativeMediaCoreProgramFramePreview? DecodeProgramFramePreview(NativeMediaCoreProgramFramePreview preview)
    {
        if (preview.Bgra is { Length: > 0 })
        {
            return preview;
        }

        if (string.IsNullOrWhiteSpace(preview.BgraBase64))
        {
            return null;
        }

        try
        {
            var bgra = Convert.FromBase64String(preview.BgraBase64);
            if (bgra.Length != preview.Width * preview.Height * 4)
            {
                return null;
            }

            return preview with { Bgra = bgra };
        }
        catch (FormatException)
        {
            return null;
        }
    }

    private static IReadOnlyList<NativeMediaCoreOutputHealth> BuildOutputHealth(
        IReadOnlyList<string> outputs,
        NativeMediaCoreOutputSenderSession? senderSession,
        NativeMediaCoreRecordingSession? recording,
        IReadOnlyList<string> encoderWarnings)
    {
        var health = new List<NativeMediaCoreOutputHealth>();
        foreach (var destination in outputs)
        {
            if (destination == "recording")
            {
                var status = recording?.Status switch
                {
                    "failed" => "failed",
                    "warning" => "warning",
                    _ when recording?.WriterStatus == "warning" => "warning",
                    _ when recording?.Active == true => "live",
                    _ => "idle"
                };
                health.Add(new NativeMediaCoreOutputHealth
                {
                    Destination = destination,
                    Status = status,
                    Message = recording?.Error ?? recording?.Warning ?? recording?.ProgramPath ?? "Recording idle.",
                    DroppedFrames = recording?.TotalDroppedFrames ?? 0
                });
                continue;
            }

            var sender = senderSession?.Senders.FirstOrDefault(item => item.Destination == destination);
            health.Add(new NativeMediaCoreOutputHealth
            {
                Destination = destination,
                Status = sender?.Status switch
                {
                    "failed" => "failed",
                    "warning" => "warning",
                    "live" or "starting" => "live",
                    _ => "idle"
                },
                Message = sender?.Warning ?? $"{destination.ToUpperInvariant()} sender {sender?.Status ?? "idle"}.",
                DroppedFrames = 0
            });
        }

        foreach (var warning in encoderWarnings)
        {
            health.Add(new NativeMediaCoreOutputHealth
            {
                Destination = "recording",
                Status = "warning",
                Message = warning,
                DroppedFrames = 0
            });
        }

        return health;
    }
}
